core/src/apps/ethereum/sign_typed_data.py

```diff
@@ -71,32 +71,46 @@ async def generate_typed_data_hash(
     )
     await typed_data_envelope.collect_types()
 
+    # EIP-712 prefix
+    parts = [b"\x19\x01"]
+
     name, version = await get_name_and_version_for_domain(ctx, typed_data_envelope)
     show_domain = await should_show_domain(ctx, name, version)
-    domain_separator = await typed_data_envelope.hash_struct(
-        primary_type="EIP712Domain",
-        member_path=[0],
-        show_data=show_domain,
-        parent_objects=["EIP712Domain"],
+    # domain_seperator
+    parts.append(
+        await typed_data_envelope.hash_struct(
+            primary_type="EIP712Domain",
+            member_path=[0],
+            show_data=show_domain,
+            parent_objects=["EIP712Domain"],
+        )
     )
 
-    show_message = await should_show_struct(
-        ctx,
-        description=primary_type,
-        data_members=typed_data_envelope.types[primary_type].members,
-        title="Confirm message",
-        button_text="Show full message",
-    )
-    message_hash = await typed_data_envelope.hash_struct(
-        primary_type=primary_type,
-        member_path=[1],
-        show_data=show_message,
-        parent_objects=[primary_type],
-    )
+    # Setting the primary_type to "EIP712Domain" is technically in spec
+    # In this case, we ignore the "message" part and only use the "domain" part
+    # https://ethereum-magicians.org/t/eip-712-standards-clarification-primarytype-as-domaintype/3286
+    if primary_type != "EIP712Domain":
+        show_message = await should_show_struct(
+            ctx,
+            description=primary_type,
+            data_members=typed_data_envelope.types[primary_type].members,
+            title="Confirm message",
+            button_text="Show full message",
+        )
+        parts.append(
+            await typed_data_envelope.hash_struct(
+                primary_type=primary_type,
+                member_path=[1],
+                show_data=show_message,
+                parent_objects=[primary_type],
+            )
+        )
+
+    full_bytes_to_hash = b"".join(parts)
 
-    await confirm_hash(ctx, message_hash)
+    await confirm_hash(ctx, full_bytes_to_hash)
```

Comment:
This way even the `\x19\x01` prefix and also the `domain_separator` will be shown on the screen, instead of just showing the `message_hash` as before. We probably do not want that (or at least it changes the previous state).

Hard to say what is the best way of doing it. We could handle the confirmation dialog in if/else depending on `primary_type == "EIP712Domain"`. In that if/else, it could be also possible to set `message_hash = b""` when we do not care about it, so we can keep the below return statement unchanged (and maybe more importantly also get rid of the array-creating-and-appending-and-joining logic)

Reply:
I didn't notice this originally, but we definitely do want the domain hash to also show. 

Showing the full thing, including `\x19\x01` makes sense.

Reply:
Personally, I'd agree with @matejcik. It makes more sense to show the `0x1901` prefix and `domain_seperator`, even though it changes the previous state. I'll add it in, but if we want to keep it using the same state, I can always revert back to the old behaviour.

Reply:
I've removed the array-creating-and-appending-and-joining logic so the code should be more similar to how it previously was, so the total lines changed should be less.

I've made the changes in a `sqaush! ...` commit in https://github.com/trezor/trezor-firmware/pull/2036/commits/5024b0165fe0f45c43af2f9032334d545bdbebd5, so that it will be put in the commit message when we `--autosquash`. It seems like a minor UI change, but I'd also be happy to put it in the CHANGELOG entry too.

I have however kept the change that shows the `0x1901` and `domain_sepeaator` on the screen. I think it makes it more clear, since now it will show the full data that is being hashed, especially since the `domain_separator` is variable data, but I'm happy to revert it if you disagree.

Screenshot below shows it still shows on one screen, so no scrolling required.
![emu00000048](https://user-images.githubusercontent.com/19716675/148307090-fe94cbcf-f45b-4717-b68d-551a41344638.png)